File: web/lib/linalg.js
// Cholesky solve for the ridge normal equations.
//
// Port of ranksim/linalg.py. The design matrices are (alliance-appearances x
// teams) with ~30 columns, so a plain O(n^3) Cholesky is instant even in a
// phone's JS engine. tests/test_parity.py asserts this agrees with the Python.

export function cholesky(a) {
  const n = a.length;
  const lower = Array.from({ length: n }, () => new Float64Array(n));
  for (let i = 0; i < n; i++) {
    for (let j = 0; j <= i; j++) {
      let s = 0;
      for (let k = 0; k < j; k++) s += lower[i][k] * lower[j][k];
      if (i === j) {
        let d = a[i][i] - s;
        // Ridge should prevent this; nudge rather than explode.
        if (d <= 0) d = 1e-9;
        lower[i][j] = Math.sqrt(d);
      } else {
        lower[i][j] = (a[i][j] - s) / lower[j][j];
      }
    }
  }
  return lower;
}

export function cholSolve(lower, b) {
  const n = lower.length;
  const y = new Float64Array(n);
  for (let i = 0; i < n; i++) {
    let s = 0;
    for (let k = 0; k < i; k++) s += lower[i][k] * y[k];
    y[i] = (b[i] - s) / lower[i][i];
  }
  const x = new Float64Array(n);
  for (let i = n - 1; i >= 0; i--) {
    let s = 0;
    for (let k = i + 1; k < n; k++) s += lower[k][i] * x[k];
    x[i] = (y[i] - s) / lower[i][i];
  }
  return x;
}

export function cholInvDiag(lower) {
  const n = lower.length;
  const diag = new Float64Array(n);
  for (let col = 0; col < n; col++) {
    const e = new Float64Array(n);
    e[col] = 1;
    diag[col] = cholSolve(lower, e)[col];
  }
  return diag;
}

// Build (X'X + ridge*I, X'y) from sparse rows of column indices. Every design
// row is an alliance: three 1s and the rest zeros.
export function normalEquations(rows, y, nCols, ridge) {
  const ata = Array.from({ length: nCols }, () => new Float64Array(nCols));
  const aty = new Float64Array(nCols);
  for (let r = 0; r < rows.length; r++) {
    const cols = rows[r];
    const target = y[r];
    for (const i of cols) {
      aty[i] += target;
      for (const j of cols) ata[i][j] += 1;
    }
  }
  for (let i = 0; i < nCols; i++) ata[i][i] += ridge;
  return { ata, aty };
}

export function mean(values) {
  if (!values.length) return 0;
  let s = 0;
  for (const v of values) s += v;
  return s / values.length;
}

export function median(values) {
  if (!values.length) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const mid = sorted.length >> 1;
  return sorted.length % 2 ? sorted[mid] : (sorted[mid - 1] + sorted[mid]) / 2;
}

File: ranksim/linalg.py
"""Tiny dense linear algebra: Cholesky solve for the ridge normal equations.

The design matrices here are (alliance-appearances x teams) with ~30 columns, so a
plain O(n^3) Cholesky in pure Python is instant and keeps the tool dependency-free.
"""

from __future__ import annotations

import math


def cholesky(a: list[list[float]]) -> list[list[float]]:
    """Lower-triangular L with L @ L.T == a, for symmetric positive definite a."""
    n = len(a)
    lower = [[0.0] * n for _ in range(n)]
    for i in range(n):
        for j in range(i + 1):
            s = sum(lower[i][k] * lower[j][k] for k in range(j))
            if i == j:
                d = a[i][i] - s
                if d <= 0.0:
                    # Ridge should prevent this; nudge rather than explode.
                    d = 1e-9
                lower[i][j] = math.sqrt(d)
            else:
                lower[i][j] = (a[i][j] - s) / lower[j][j]
    return lower


def chol_solve(lower: list[list[float]], b: list[float]) -> list[float]:
    """Solve (L L.T) x = b."""
    n = len(lower)
    y = [0.0] * n
    for i in range(n):
        y[i] = (b[i] - sum(lower[i][k] * y[k] for k in range(i))) / lower[i][i]
    x = [0.0] * n
    for i in range(n - 1, -1, -1):
        x[i] = (y[i] - sum(lower[k][i] * x[k] for k in range(i + 1, n))) / lower[i][i]
    return x


def chol_inv_diag(lower: list[list[float]]) -> list[float]:
    """Diagonal of (L L.T)^-1, used for parameter standard errors."""
    n = len(lower)
    diag = [0.0] * n
    for col in range(n):
        e = [0.0] * n
        e[col] = 1.0
        diag[col] = chol_solve(lower, e)[col]
    return diag


def normal_equations(
    rows: list[list[int]], y: list[float], n_cols: int, ridge: float
) -> tuple[list[list[float]], list[float]]:
    """Build (X'X + ridge*I, X'y) from sparse rows of column indices.

    Every design row here is an alliance: three 1s and the rest zeros, so the
    matrix is assembled from index pairs instead of a dense multiply.
    """
    ata = [[0.0] * n_cols for _ in range(n_cols)]
    aty = [0.0] * n_cols
    for cols, target in zip(rows, y):
        for i in cols:
            aty[i] += target
            for j in cols:
                ata[i][j] += 1.0
    for i in range(n_cols):
        ata[i][i] += ridge
    return ata, aty
